File: VB/MultiDateEditInfoArgs.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports DevExpress.XtraEditors.ViewInfo
Imports System.Drawing

Namespace MultiDateEditor
	Friend Class MultiDateEditInfoArgs
		Inherits DateEditInfoArgs
		' constructor
		Public Sub New(ByVal OwnerCalendar As MultiDateEditCalendar)
			MyBase.New(OwnerCalendar)
		End Sub

		Protected Sub SetMonthYearButtonState(ByVal DMState As Boolean, ByVal IMState As Boolean, ByVal DYState As Boolean, ByVal IYState As Boolean)
			ShowDecMonth = DMState
			ShowIncMonth = IMState
			ShowDecYear = DYState
			ShowIncYear = IYState
			DecMonthInfo.State = If(DMState = True, DevExpress.Utils.Drawing.ObjectState.Normal, DevExpress.Utils.Drawing.ObjectState.Disabled)
			IncMonthInfo.State = If(IMState = True, DevExpress.Utils.Drawing.ObjectState.Normal, DevExpress.Utils.Drawing.ObjectState.Disabled)
			DecYearInfo.State = If(DYState = True, DevExpress.Utils.Drawing.ObjectState.Normal, DevExpress.Utils.Drawing.ObjectState.Disabled)
			IncYearInfo.State = If(IYState = True, DevExpress.Utils.Drawing.ObjectState.Normal, DevExpress.Utils.Drawing.ObjectState.Disabled)
		End Sub

		Protected Overrides Sub CalcHeaderInfo()
			MyBase.CalcHeaderInfo()
			Dim currentCalendar As MultiDateEditCalendar = TryCast(Calendar, MultiDateEditCalendar)
			Dim currentRepository As RepositoryItemMultiDateEdit = TryCast(currentCalendar.Properties, RepositoryItemMultiDateEdit)
			If currentRepository.MultiEditorType <> MultiDateEditOptions.Single Then
				If Bounds.X = currentCalendar.CalendarsRecr.Left Then ' its a first calendar
					SetMonthYearButtonState(True, False, False, False)
					EditYear = Rectangle.Empty
					EditMonth.X = Bounds.Width - EditMonth.Width - Bounds.X
					DecMonthInfo.Bounds = New Rectangle(EditMonth.X - DecMonthInfo.Bounds.Width - Bounds.X, DecMonthInfo.Bounds.Y, DecMonthInfo.Bounds.Width, DecMonthInfo.Bounds.Height)
				ElseIf (Bounds.X + Bounds.Width) = currentCalendar.CalendarsRecr.Right Then ' its a last calendar
					SetMonthYearButtonState(False, True, True, True)
					DecYearInfo.Bounds = New Rectangle(EditYear.X - IncYearInfo.Bounds.Width, IncYearInfo.Bounds.Y, IncYearInfo.Bounds.Width, IncYearInfo.Bounds.Height)

				Else
					SetMonthYearButtonState(False, False, False, False)
					EditMonth.X = Bounds.X + (Bounds.Width - EditMonth.Width) / 2
					EditYear = Rectangle.Empty
				End If
			End If
		End Sub
	End Class
End Namespace

File: CS/MultiDateEditInfoArgs.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.XtraEditors.ViewInfo;
using System.Drawing;

namespace MultiDateEditor
{
    class MultiDateEditInfoArgs : DateEditInfoArgs
    {
        // constructor
        public MultiDateEditInfoArgs(MultiDateEditCalendar OwnerCalendar) : base(OwnerCalendar) { }

        protected void SetMonthYearButtonState(bool DMState, bool IMState, bool DYState, bool IYState)
        {
            ShowDecMonth = DMState;
            ShowIncMonth = IMState;
            ShowDecYear = DYState;
            ShowIncYear = IYState;
            DecMonthInfo.State = DMState == true ? DevExpress.Utils.Drawing.ObjectState.Normal : DevExpress.Utils.Drawing.ObjectState.Disabled;
            IncMonthInfo.State = IMState == true ? DevExpress.Utils.Drawing.ObjectState.Normal : DevExpress.Utils.Drawing.ObjectState.Disabled;
            DecYearInfo.State = DYState == true ? DevExpress.Utils.Drawing.ObjectState.Normal : DevExpress.Utils.Drawing.ObjectState.Disabled;
            IncYearInfo.State = IYState == true ? DevExpress.Utils.Drawing.ObjectState.Normal : DevExpress.Utils.Drawing.ObjectState.Disabled;
        }

        protected override void CalcHeaderInfo()
        {
            base.CalcHeaderInfo();
            MultiDateEditCalendar currentCalendar = Calendar as MultiDateEditCalendar;
            RepositoryItemMultiDateEdit currentRepository = currentCalendar.Properties as RepositoryItemMultiDateEdit;
            if (currentRepository.MultiEditorType != MultiDateEditOptions.Single)
            {
                if (Bounds.X == currentCalendar.CalendarsRecr.Left) // its a first calendar
                {
                    SetMonthYearButtonState(true, false, false, false);
                    EditYear = Rectangle.Empty;
                    EditMonth.X = Bounds.Width - EditMonth.Width - Bounds.X;
                    DecMonthInfo.Bounds = new Rectangle(EditMonth.X - DecMonthInfo.Bounds.Width - Bounds.X, DecMonthInfo.Bounds.Y, DecMonthInfo.Bounds.Width, DecMonthInfo.Bounds.Height);
                }
                else if ((Bounds.X + Bounds.Width) == currentCalendar.CalendarsRecr.Right) // its a last calendar
                {
                    SetMonthYearButtonState(false, true, true, true);
                    DecYearInfo.Bounds = new Rectangle(EditYear.X - IncYearInfo.Bounds.Width, IncYearInfo.Bounds.Y, IncYearInfo.Bounds.Width, IncYearInfo.Bounds.Height);

                }
                else
                {
                    SetMonthYearButtonState(false, false, false, false);
                    EditMonth.X = Bounds.X + (Bounds.Width - EditMonth.Width) / 2;
                    EditYear = Rectangle.Empty;
                }
            }
        }
    }
}
